test/java/integration/AndroidEngineSocketTagTest.java

```diff
@@ -0,0 +1,188 @@
+import io.envoyproxy.envoymobile.FinalStreamIntel;
+import io.envoyproxy.envoymobile.LogLevel;
+import io.envoyproxy.envoymobile.RequestHeaders;
+import io.envoyproxy.envoymobile.RequestHeadersBuilder;
+import io.envoyproxy.envoymobile.RequestMethod;
+import io.envoyproxy.envoymobile.ResponseHeaders;
+import io.envoyproxy.envoymobile.ResponseTrailers;
+import io.envoyproxy.envoymobile.Stream;
+import io.envoyproxy.envoymobile.StreamIntel;
+import io.envoyproxy.envoymobile.UpstreamHttpProtocol;
+import io.envoyproxy.envoymobile.engine.AndroidJniLibrary;
+import io.envoyproxy.envoymobile.engine.testing.RequestScenario;
+import io.envoyproxy.envoymobile.engine.testing.Response;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class AndroidEngineSocketTagTest {
+
+  private final MockWebServer mockWebServer = new MockWebServer();
+  private Engine engine;
+
+  @BeforeClass
+  public static void loadJniLibrary() {
+    AndroidJniLibrary.loadTestLibrary();
+  }
+
+  @Before
+  public void setUpEngine() throws Exception {
+    CountDownLatch latch = new CountDownLatch(1);
+    Context appContext = ApplicationProvider.getApplicationContext();
+    engine = new AndroidEngineBuilder(appContext)
+                 .addLogLevel(LogLevel.OFF)
+                 .enableSocketTagging(true)
+                 .setOnEngineRunning(() -> {
+                   latch.countDown();
+                   return null;
+                 })
+                 .build();
+    latch.await(); // Don't launch a request before initialization has completed.
+  }
+
+  @After
+  public void shutdownEngine() throws Exception {
+    engine.terminate();
+    mockWebServer.shutdown();
+  }
+
+  @Test
+  public void socket_tag() throws Exception {
+    mockWebServer.setDispatcher(new Dispatcher() {
+      @Override
+      public MockResponse dispatch(RecordedRequest recordedRequest) {
+        assertThat(recordedRequest.getMethod()).isEqualTo(RequestMethod.GET.name());
+        assertThat(recordedRequest.getHeader("x-envoy-mobile-socket-tag")).isEqualTo(null);
```

Comment:
why is it expected to be equal to `null` if we set a valid socket tag on the request?

Reply:
ah, because we remove it when we detect that it's present

Reply:
Yup, exactly!

Reply:
Any way we could ensure that we are really adding `x-envoy-mobile-socket-tag` at some point? As it is now this test is always green - even if socket tagging is not enabled / does not end up adding a tag.

Reply:
Filed https://github.com/envoyproxy/envoy-mobile/issues/2494 as a result of offline discussions. Thanks for working through this!